Migration doc for GlobalSettings

dotta opened the pull request:
Migration doc for helping users moving away from `GlobalSettings` in the upcoming Play 2.4 release.

Note that some of the links I used are broken (missing page). For instance, we don't have a `JavaHttpRequestHandlers` page in the doc. I've worked under the assumption that we will create the missing pages before 2.4 final is published. Let me know if I'm wrong.

dotta commented on documentation/manual/releases/Migration24.md:
```diff
@@ -152,6 +153,61 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+
+* `GlobalSettings.onRequest`: Create a class that extends [`DefaultHttpRequestHandler`](api/java/play/http/DefaultHttpRequestHandler.html), and move the implementation of your `GlobalSettings.onRequest` method inside the `DefaultHttpRequestHandler.createAction` method. Read [[Request Handlers|JavaHttpRequestHandlers]] for more information.
+
+* `GlobalSettings.onRouteRequest`: FIXME: I don't see how this method can be refactored. `JavaCompatibleHttpRequestHandler.routeRequest` is involved, but the injected `JavaCompatibleHttpRequestHandler` instance cannot be replaced. I need help on this one.
```
I need help on this one /cc @jroper @richdougherty

jroper replied:
Implement a custom router.  This method leaks the Scala API a bit, and is not easy, or even really possible to implement in Java, which is why I've left it out.

dotta replied:
I see what you mean. I feel slightly bad because Java here feels 2nd class. I understand the technical reasons, and I know we are working to remove these limitations in the future. I guess the best thing to do here is simply hinting that there is no easy migration for this method.

benmccann commented on documentation/manual/releases/Migration24.md:
```diff
@@ -152,6 +153,61 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+
+* `GlobalSettings.onStop`: Add a dependency to [`ApplicationLifecycle`](api/java/play/inject/ApplicationLifecycle.html) on the class that needs to register a stop hook. Then, move the implementation of your `GlobalSettings.onStop` method inside the `Promise` passed to the `ApplicationLifecycle.addStopHook`. Read [[Stopping/cleaning-up|JavaDependencyInjection#Stopping/cleaning-up]] for more inform ...
+
+* `GlobalSettings.onError`: Create a class that extends [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and move the implementation of your `GlobalSettings.onError` inside the `HttpErrorHandler.onServerError` method. Read [[Error Handling|JavaErrorHandling]] for more information.
```
This should be "implements" instead of "extends"

dotta replied:
It's a good point. I'm guessing you are suggesting to make the change only for the Java doc, when a class need to `implements` an interface. However, your suggestion made me think that I'd like to use a broader, language independent, term (since `extends` is not accurate in Scala either). So, I'm going to chance "extends" with "inherits from". WDYT?

dotta replied:
Using "inherits from" is also more accurate because you may end up extending `DefaultHttpRequestHandler` instead of `HttpErrorHandler`.

benmccann commented on documentation/manual/releases/Migration24.md:
```diff
@@ -152,6 +153,61 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+
+* `GlobalSettings.onRouteRequest`: FIXME: I don't see how this method can be refactored. `JavaCompatibleHttpRequestHandler.routeRequest` is involved, but the injected `JavaCompatibleHttpRequestHandler` instance cannot be replaced. I need help on this one.
+
+* `GlobalSettings.onHandlerNotFound`: Create a class that extends [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide and implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
```
What status codes should you check for? I.E. how do you differentiate between `onHandlerNotFound` and `onBadRequest`?

jroper replied:
`onClientError` will be invoked with the status code as an argument... ie, 400 will be passed for bad request, 404 for handler not found.

dotta replied:
Yep, that is why I said "provide an implementation" (and just noticed there is a typo! :-)). I'll see how I can reword it to hint to use the status code.

dotta commented:
@play-pull-request-validator

cchantep commented:
Maybe the #4308 can be covered by additional changes in this PR.

jroper commented:
@dotta What was your plan for this?  We can't merge with broken links, since that will break the build.

Also, a lot of the links to Scala documentation are broken, they should be in the format `/api/scala/index.html#play.api.Foo`, not `/api/scala/play/api/Foo.html`.

dotta commented:
@jroper If merging this PR will break the build, then we definitely shouldn't merge it. I'll work on the missing doc that this PR requires so that we will can merge this soon. Also, thanks for the feedback on the Scala doc, I'll fix it.

Is it ok if this PR stays open (so that we don't forget about it). Or should it be closed until it's ready to merge?

jroper commented:
Keep open until ready for merge.

richdougherty commented:
If you want to do the docs in multiple PRs that's fine too. We can merge the bits that are ready and then have more PRs when you write more docs. Completely up to you how you want to do it though. :)

jroper commented:
As long as the build passes - our tests test that there are no dead internal links :)

richdougherty commented:
@play-pull-request-validator, please build again now that link validation for scaladoc has changed.

richdougherty commented:
That's a weird error in NingConfigSpec:

```
[info] NingConfigSpec
[info] 
[info] NingConfigSpec should
[error]   ! case class defaults must match reference.conf defaults
[error]    null (PropertiesParser.java:60)
[error] com.typesafe.config.impl.PropertiesParser.fromProperties(PropertiesParser.java:60)
[error] com.typesafe.config.impl.Parseable$ParseableProperties.rawParseValue(Parseable.java:860)
[error] com.typesafe.config.impl.Parseable$ParseableProperties.rawParseValue(Parseable.java:842)
[error] com.typesafe.config.impl.Parseable.parseValue(Parseable.java:180)
[error] com.typesafe.config.impl.Parseable.parseValue(Parseable.java:174)
[error] com.typesafe.config.impl.Parseable.parse(Parseable.java:299)
[error] com.typesafe.config.ConfigFactory.parseProperties(ConfigFactory.java:638)
[error] com.typesafe.config.ConfigFactory.parseProperties(ConfigFactory.java:649)
[error] play.api.Configuration$.load(Configuration.scala:44)
[error] play.api.Configuration$.load(Configuration.scala:123)
[error] play.api.Configuration$.load(Configuration.scala:129)
[error] play.api.inject.guice.GuiceApplicationBuilder$$anonfun$$lessinit$greater$default$6$1.apply(GuiceApplicationBuilder.scala:19)
[error] play.api.inject.guice.GuiceApplicationBuilder$$anonfun$$lessinit$greater$default$6$1.apply(GuiceApplicationBuilder.scala:19)
[error] play.api.inject.guice.GuiceApplicationBuilder.injector(GuiceApplicationBuilder.scala:66)
[error] play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:93)
[error] play.api.test.FakeApplication.<init>(Fakes.scala:216)
[error] play.api.test.WithApplication$.$lessinit$greater$default$1(Specs.scala:36)
[error] play.api.libs.ws.ning.NingConfigSpec$$anonfun$6$$anonfun$apply$34$$anon$1.<init>(NingConfigSpec.scala:44)
[error] play.api.libs.ws.ning.NingConfigSpec$$anonfun$6$$anonfun$apply$34.apply(NingConfigSpec.scala:44)
[error] play.api.libs.ws.ning.NingConfigSpec$$anonfun$6$$anonfun$apply$34.apply(NingConfigSpec.scala:44)
```

@play-pull-request-validator, please try again.

jroper commented:
@play-pull-request-validator

richdougherty commented:
Closing because of #4458.

jroper commented:
@dotta This still has invalid links wiki links, and so can't be merged:

https://playframework2.ci.cloudbees.com/job/play-master-PRs/4629/console

dotta commented:
I've rebase against 2.4.x, but maybe I should have rebased against master(?). Not sure what's the standard way of doing it in Play. Let me know.

richdougherty commented:
Usually you merge into _master_ then `cherry-pick` onto the `2.4.x` branch. If you don't want to `cherry-pick` directly onto a branch (or don't have access) you can create a second PR against the branch once the first PR has been merged.

dotta commented:
@richdougherty Thanks Rich. By the way, this PR is (finally!) up for review :+1:

richdougherty commented on documentation/manual/releases/Migration24.md:
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+
+#### Java
+
+* `GlobalSettings.beforeStart` and `GlobalSettings.onStart`:  Anything that needs to happen on start up should now be happening in the constructor of a dependency injected class. A class will perform its initialisation when the dependency injection framework loads it. If you need eager initialisation (for example, because you need to execute some code *before* the application is actually started) ...
```
There are now docs on eager bindings: https://www.playframework.com/documentation/2.4.0-RC3/JavaDependencyInjection#Eager-bindings

richdougherty commented on documentation/manual/releases/Migration24.md:
````diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+* `GlobalSettings.onHandlerNotFound`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide an implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```scala
````
Should this be Java code?

richdougherty commented on documentation/manual/releases/Migration24.md:
````diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+* `GlobalSettings.onBadRequest`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide an implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```scala
````
This should be Java code too, I think.

richdougherty commented on documentation/manual/releases/Migration24.md:
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+
+#### Scala
+
+* `GlobalSettings.beforeStart` and `GlobalSettings.onStart`:  Anything that needs to happen on start up should now be happening in the constructor of a dependency injected class. A class will perform its initialisation when the dependency injection framework loads it. If you need eager initialisation (because you need to execute some code *before* the application is actually started), use the `ea ...
```
Link to docs: https://www.playframework.com/documentation/2.4.0-RC3/ScalaDependencyInjection#Eager-bindings

richdougherty commented on documentation/manual/releases/Migration24.md:
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+
+* `GlobalSettings.onRequest`: Create a class that inherits from [`DefaultHttpRequestHandler`](api/java/play/http/DefaultHttpRequestHandler.html), and move the implementation of your `GlobalSettings.onRequest` method inside the `DefaultHttpRequestHandler.createAction` method. Read [[Request Handlers|JavaHttpRequestHandlers]] for more information.
+
+* `GlobalSettings.onRouteRequest`: There is no simple migration for this method when using the Java API. If you need this, you will have to keep your Global class around for a little longer.
```
Can they extend `HttpRequestHandler` or is that only in Scala?

dotta replied:
That's only valid in Scala :-( see https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/http/HttpRequestHandler.scala#L42

richdougherty commented on documentation/manual/working/javaGuide/advanced/http/JavaHttpRequestHandlers.md:
```diff
@@ -0,0 +1,31 @@
+The [`HttpRequestHandler`](api/java/play/http/HttpRequestHandler) interface has two methods that needs to be implemented: 
+
+* `createAction`: Takes the request and the controller's action method associated with the passed request.
+*  `wrapAction`: Takes the action to be run and allows for a final global intereceptor to be added to the action.
```
s/inter**e**ceptor/interceptor/

richdougherty commented:
Great work @dotta! A few more changes, but nearly there now…

dotta commented:
@richdougherty I've addressed all your comments (by the way, thanks!). Let me know if it looks good now, so that I can squash commits (as it makes sense), and we can finally merge this baby :-)

richdougherty commented on documentation/manual/releases/Migration24.md:
````diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```java
+if(statusCode == play.api.http.Status.NOT_FOUND) {
````
`play.api` is the Scala API, so I think this should be `play.mvc.Http.Status.NOT_FOUND`.

dotta replied:
I didn't realise we have two places with the constants. I'll make the change, but I wonder why keeping things this way. Would it be bad to deprecate `play.api.http.Status`?

One more thing, it looks like that  `play.api.http.Status` and `play.mvc.Http.Status` don't actually contain exactly the same constants. Is this an oversight due to maintaining the same bits in two different places?

dotta replied:
All good now. No need to comment.

richdougherty commented on documentation/manual/releases/Migration24.md:
````diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```java
+if(statusCode == play.api.http.Status.BAD_REQUEST) {
````
Same here: `play.mvc.Http.Status.BAD_REQUEST`.

richdougherty commented:
@dotta, just two small changes, then if you squash the commits I'd be happy to merge this. :)

dotta commented:
@richdougherty All changes done. I've also rebased against master and squashed commits as it made sense. Feel free to merge when the PR validator has finished running. By the way, could you have a look at [this comment](https://github.com/playframework/playframework/pull/4308#discussion_r30570045) and let me know what you think.

dotta commented:
Alright, finally ready to merge, together with the [backport for 2.4.x](https://github.com/playframework/playframework/pull/4500)
